neat_pvd.c

```diff
@@ -0,0 +1,531 @@
+#ifdef HAVE_SIN_LEN
+        async_query->dst_addr4->sin_len     = sizeof(struct sockaddr_in);
+#endif
+
+        if (uv_udp_send(async_query->dns_snd_handle,
+                        async_query->resolve_handle,
+                        async_query->dns_uv_snd_buf,
+                        1,
+                        (const struct sockaddr*) async_query->dst_addr4,
+                        send_cb)) {
+            neat_log(NEAT_LOG_ERROR, "%s - Failed to start DNS send", __func__);
+            return 1;
+        }
+    } else {
+        server_addr6                        = (struct sockaddr_in6 *) dns_addr;
+        async_query->dst_addr6              = calloc(sizeof(struct sockaddr_in6), 1);
+        async_query->dst_addr6->sin6_family = AF_INET6;
+        async_query->dst_addr6->sin6_port   = htons(LDNS_PORT);
+        async_query->dst_addr6->sin6_addr   = server_addr6->sin6_addr;
+#ifdef HAVE_SIN6_LEN
+        async_query->dst_addr6->sin6_len    = sizeof(struct sockaddr_in6);
+#endif
+
+        if (uv_udp_send(async_query->dns_snd_handle,
+                        async_query->resolve_handle,
+                        async_query->dns_uv_snd_buf,
+                        1,
+                        (const struct sockaddr*) async_query->dst_addr6,
+                        send_cb)) {
+            neat_log(NEAT_LOG_ERROR, "%s - Failed to start DNS send", __func__);
+            return 1;
+        }
+    }
+
+    return 0;
+}
+
+//Called when a DNS request has been (i.e., passed to socket). We will send the
+//second query (used for checking poisoning) here. If that is needed
+static void
+neat_pvd_dns_sent_cb(uv_udp_send_t *req, int status)
+{
+}
+
+static void
+neat_pvd_free_async_query(struct pvd_async_query *async_query)
+{
+    uv_udp_recv_stop(async_query->resolve_handle);
+    free(async_query->dns_uv_snd_buf);
+    free(async_query->dns_snd_handle);
+    ldns_buffer_free(async_query->dns_snd_buf);
+    if (async_query->dst_addr4 != NULL)
+        free(async_query->dst_addr4);
+    if (async_query->dst_addr6 != NULL)
+        free(async_query->dst_addr6);
+    free(async_query->resolve_handle);
+
+    LIST_REMOVE(async_query, next_query);
+    free(async_query);
+}
+
+//This callback is called when we close a UDP socket (handle) and allows us to
+//free any allocated resource. In our case, this is only the dns_snd_buf
+static void
+neat_pvd_dns_close_cb(uv_handle_t *handle)
+{
+    struct pvd_async_query *async_query = handle->data;
+
+    neat_pvd_free_async_query(async_query);
+}
+
+//libuv gives the user control of how memory is allocated. This callback is
+//called when a UDP packet is ready to received, and we have to fill out the
+//provided buf with the storage location (and available size)
+static void
+neat_pvd_dns_alloc_cb(uv_handle_t *handle,
+                      size_t suggested_size,
+                      uv_buf_t *buf)
+{
+    char *dns_rcv_buf = calloc(sizeof(char), 1472);
```

Comment:
1472 seems like a magic number. Is it based on anything specific or is it an estimate?